source/xcb.c

```diff
@@ -1346,7 +1350,14 @@ static gboolean main_loop_x11_event_handler(xcb_generic_event_t *ev,
     // status);
     return G_SOURCE_CONTINUE;
   }
+
   uint8_t type = ev->response_type & ~0x80;
+  if (!xcb_xim_filter_event(xcb->im, ev) && xcb->ic &&
+      (type == XCB_KEY_PRESS)) {
```

Comment:
I feel like in some cases, the input-method path would not trigger, and the flat removal of `XCB_KEY_PRESS` handling would be missing  
Also, I didn’t find the doc of `xcb_xim_filter_event()`, but it sounds like you don’t really want it to be hidden in a multiple-conditions thing

I’d rather have:
* the `xcb_xim_filter_event()` call in its own `if` with an early `return` to actually filter xim events **and** with a prefixed `xcb->im != NULL` check, unless you can show me it’ll never be `NULL`, for anyone, ever
* the `xcb->ic != NULL` in the existing `XCB_KEY_PRESS`, to branch the IM vs non-IM behaviours
* the current `XCB_KEY_PRESS` logic in a separate function but **not** the xim callback (to keep only the necessary args in it)

Reply:
> Also, I didn’t find the doc of xcb_xim_filter_event(), but it sounds like you don’t really want it to be hidden in a multiple-conditions thing

documentation is lacking, but apparently we want to filter all events with xcb_xim_filter, and only forward when them when the ic is created (xcb->ic). 
https://github.com/fcitx/xcb-imdkit/issues/2#issuecomment-362765322

> xcb->im != NULL check, unless you can show me it’ll never be NULL, for anyone

from my tests, im was never null, think it's just a container to communicate with the xim server we create on the libgwater part, what defines if there's an IM server or not is by checking if the ic was created (xcb->ic), which I'll fix on the commit.

but I guess it doesn't hurt to check just to be sure, when I get the time to look through the code, and confirm it, we can remove those checks.